README.md
Halfred
=======

[![Build Status](https://travis-ci.org/basti1302/halfred.png?branch=master)](https://travis-ci.org/basti1302/halfred)

[![browser support](http://ci.testling.com/basti1302/halfred.png)](http://ci.testling.com/basti1302/halfred)

[![NPM](https://nodei.co/npm/halfred.png?downloads=true&stars=true)](https://nodei.co/npm/halfred/)

A "parser" for the JSON-flavour of HAL, the Hypertext Application Language (that is `application/hal+json`). If you feed it an object that has `_links` and `_embedded` properties, as desribed in the HAL spec, it will make all links and embedded resource available via convenient methods. If requested, Halfred can also validate a HAL object.

For more information on HAL, see
* [the formal spec](http://tools.ietf.org/html/draft-kelly-json-hal)
* [a less formal introduction](http://stateless.co/hal_specification.html)

This module works in Node.js and in the browser. It has no dependencies, the size of the browser build is 8 KB / 4 KB (non-minified/minified)

Installation
------------

### Node.js

npm:

    npm install halfred --save

### Browser

Download and use one of the following:

* [Minified build with UMD](https://raw.github.com/basti1302/halfred/master/browser/dist/halfred.min.js): This build can be used with an AMD loader like RequireJS or with a script tag (in which case it will register `halfred` in the global scope). **If in doubt, use this build.**
* [Non-minified build with UMD](https://raw.github.com/basti1302/halfred/master/browser/dist/halfred.js): Same as above, just larger :-)
* If your project uses browserify, you don't need to do anything. Just `npm install halfred --save` and `require('halfred')`, then browserify your module as usual.
* [Minified require/external build](https://raw.github.com/basti1302/halfred/master/browser/dist/halfred.external.min.js): Created with browserify's `--require` parameter and intended to be used (required) from other browserified modules, which were created with `--external halfred`. This build could be used if you use browserify but do not want to bundle Halfred with your own browserify build but keep it as a separate file.
* [Non-minified require/external build](https://raw.github.com/basti1302/halfred/master/browser/dist/halfred.external.js): Same as before, just bigger.

### Usage

    var halfred = require('halfred');
    var resource = halfred.parse(object);

#### Resource API

`halfred.parse(object)` returns a resource object. Here's what you can do with it:

* `allLinkArrays()`: Returns an object which has an array for each link that was present in the source object. See below why each link is represented as an array.
* `allLinks()`: Alias for `allLinkArrays`
* `linkArray(key)`: Returns the array of links for the given `key`, or `null` if there are no links for this `key`.
* `link(key)`: Returns the first element of the array of links for the given `key` or `null` if there are no links for this `key`.
* `allEmbeddedResourceArrays()`: Returns an object which has an array for each embedded resource that was present in the source object. See below why each embedded resource is represented as an array. Each element of any of this arrays is in turn a `Resource` object.
* `allEmbeddedArrays()`: Alias for `allEmbeddedResourceArrays()`
* `allEmbeddedResources()`: Alias for `allEmbeddedResourceArrays()`
* `embeddedResourceArray(key)`:  Returns the array of embedded resources for the given `key`, or `null` if there are no embedded resources for this `key`. Each element of this arrays is in turn a `Resource` object.
* `embeddedArray(key)`: Alias for `embeddedResourceArray()`.
* `embeddedResource(key)`: Returns the first element of the array of embedded resources for the given `key` or `null` if there are no embedded resources for this `key`. The returend object is a `Resource` object.
* `embedded(key)`: Alias for `embeddedResource(key)`
* `original()`: Returns the unmodified, original object that was parsed to this resource. This is rather uninteresting for the source object you give to the `parse` method (because you probably still have a reference to the source object) but it is a convenient way to get the part of the source object that corresponds to an embedded resource.
* `curieArray`: Returns the array of curies. Each object in the array is a link object, which means it can be templated etc. See below for link object API.
* `curie(name)`: Returns the curie with the given name, if any. The returned object is a link object, which means it can be templated etc. See below for link object API.
* `validationIssues()`: Returns all validation issues. Validation issues are only gathered if validation has been turned on by calling `halfred.enableValidation()` before calling `halfred.parse`.
* `validation()`: Alias for `validationIssues()`

In addition to the methods mentioned here, `resource` has all properties of the source object. This is also true for embedded resource objects. The non-HAL properties (that is, any property except `_links` and `_embedded`) are copied over to the Resource object. This is always a shallow copy, so modifying the a non-HAL property in the Resource object might also alter the source object and vice versa.

The Resource object also has the properties `_links` and `_embedded` but they might differ from the `_links`/`_embedded` properties in the source object (Halfred applies some normalization to them). These are not intended to be accessed by clients directly, instead, use the provided methods to work with links and embedded resources.

#### Links And Embedded Resources

The resource methods `allLinkArrays()` and `linkArray(key)` an array for each link, instead of a single object. This might seem counterintuitive. The HAL spec allows a link to be either a single link object or an array of link objects, so halfred normalizes all that are not arrays to be single element arrays. If you are sure that there is only one link for a given key, you can use `link(key)` to directly get the first element from that array.

The same is true for embedded resources.

Once you have the link object, you can access the properties `href`, `templated` and so on (refer to the [spec](http://tools.ietf.org/html/draft-kelly-json-hal) for details) on it. The `templated` property defaults to false, if it wasn't set in the object given to `parse`, all other properties described in the spec default to `null`.

#### Enable/Disable Validation

In some situations, it might be desirable to validate the resource you want to parse and check, if it is valid according to the HAL spec. By default, Halfred does not do validation checks. If you want to have validation checks you can enable them by calling `halfred.enableValidation()`. Then after parsing a source object, call `validationIssues()` on the resource object returned by `parse` to get an array of all validation issues.

 You can disable validation checks again by calling `halfred.disableValidation()`.You can also call `halfred.enableValidation(true)` or `halfred.enableValidation(false)` to enable/disable validation.

Release Notes
-------------

* 0.3.0 2014-06-24: Parse curies
* 0.2.0 2013-11-22: Make the source object of a parsed resource available (useful for embedded resources)
* 0.1.1 2013-11-21: Leave source object untouched while parsing
* 0.1.0 2013-11-21: Initial release

License
-------

MIT


CHANGELOG.md
# [2.0.0](https://github.com/traverson/halfred/compare/v1.1.1...v2.0.0) (2020-06-07)


### Continuous Integration

* **node-versions:** updated tested node versions to match currently supported LTS versions ([1fcb3fb](https://github.com/traverson/halfred/commit/1fcb3fb12c915f6b420ec9eb090cc30634c80737))


### BREAKING CHANGES

* **node-versions:** dropped support for node versions that have reached EOL

## [1.1.1](https://github.com/traverson/halfred/compare/v1.1.0...v1.1.1) (2018-09-23)


### Bug Fixes

* **readme:** made a meaningless change to trigger a release ([1828b54](https://github.com/traverson/halfred/commit/1828b54))

Release Notes
-------------
* 0.1.0 2013-11-21: Initial release
* 0.1.1 2013-11-21: Leave source object untouched while parsing
* 0.2.0 2013-11-22: Make the source object of a parsed resource available (useful for embedded resources)
* 0.3.0 2014-06-27: Parse curies
* 0.4.0 2014-12-05: Check if link object is an actual object or a primitive
* 1.0.0 2015-05-21: Export Resource object for extension (thanks to @joshuajabbour)
* 1.1.0 2018-09-09: Add 'injectLogger' function


test/parser.js
'use strict';

var halfred = require('../halfred.js');
var fixtures = require('./fixtures');
var chai = require('chai');
var expect = chai.expect;
var assert = chai.assert;
var recordedLogs =
 { log: [],
   debug: [],
   info: [],
   warn: [],
   error: []
 };
var logger =
  { log: function(str) { recordedLogs.log.push(str); },
    debug: function(str) { recordedLogs.debug.push(str); },
    info: function(str) { recordedLogs.info.push(str); },
    warn: function(str) {
      recordedLogs.warn.push(str);
    },
    error: function(str) { recordedLogs.error.push(str); }
  };

describe('Parsing HAL', function() {

  beforeEach(function() {
    recordedLogs.log = [];
    recordedLogs.debug = [];
    recordedLogs.info = [];
    recordedLogs.warn = [];
    recordedLogs.error = [];
    halfred.injectLogger(logger);
  });

  afterEach(function() {
    halfred.disableValidation();
  });

  it('should parse the minimal example fixture', function() {
    var unparsed = fixtures.minimal.get();
    var resource = halfred.parse(unparsed);

    expect(resource.allLinkArrays()).to.exist;
    expect(resource.allLinkArrays()).to.be.an('object');
    expect(resource.allLinkArrays().self[0]).to.exist;
    expect(resource.allLinkArrays().self[0].href).to.exist;
    expect(resource.allLinkArrays().self[0].href).to.equal('dummy');
    expect(resource.linkArray('self')).to.exist;
    expect(resource.linkArray('self')).to.be.an('array');
    expect(resource.linkArray('self')[0]).to.exist;
    expect(resource.linkArray('self')[0].href).to.exist;
    expect(resource.linkArray('self')[0].href).to.equal('dummy');
    expect(resource.link('self')).to.exist;
    expect(resource.link('self')).to.be.an('object');
    expect(resource.link('self').href).to.exist;
    expect(resource.link('self').href).to.equal('dummy');
  });

  it('should not modify the source object', function() {
    var unparsed = fixtures.minimal.get();
    var resource = halfred.parse(unparsed);

    expect(unparsed._links.self).to.not.be.an('array');
    expect(unparsed._links.self).to.be.an('object');
    expect(unparsed._links.self.href).to.exist;
    expect(unparsed._links.self.templated).to.not.exist;
    expect(unparsed._embedded).to.not.exist;
    expect(unparsed._validation).to.not.exist;

    expect(resource._links.self).to.be.an('array');
    expect(resource._links.self).to.not.be.an('object');
    expect(resource._links.self[0].templated).to.exist;
    expect(resource._links.self[0].templated).to.be.false;
    expect(resource._embedded).to.exist;
    expect(resource._embedded).to.be.an('object');
    expect(resource._validation).to.exist;
    expect(resource._validation).to.be.an('array');
  });

  it('should parse the standard HAL example - main links', function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);

    expect(resource.allLinkArrays()).to.exist;
    expect(resource.link('self').href).to.equal('/orders');
    expect(resource.link('self').templated).to.be.false;
    expect(resource.link('next').href).to.equal('/orders?page=2');
    expect(resource.link('find').href).to.equal('/orders{?id}');
    expect(resource.link('find').templated).to.be.true;
  });

  it('should parse the standard HAL example - admin links', function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);

    expect(resource.allLinkArrays()).to.exist;
    expect(resource.linkArray('admin')).to.exist;
    expect(resource.linkArray('admin')).to.be.an('array');
    expect(resource.linkArray('admin').length).to.equal(2);
    expect(resource.linkArray('admin')[0].href).to.equal('/admins/2');
    expect(resource.linkArray('admin')[0].title).to.equal('Fred');
    expect(resource.linkArray('admin')[1].href).to.equal('/admins/5');
    expect(resource.linkArray('admin')[1].title).to.equal('Kate');
    expect(resource.link('admin').href).to.equal('/admins/2');
    expect(resource.link('admin').title).to.equal('Fred');
    expect(resource.link('admin', 0).href).to.equal('/admins/2');
    expect(resource.link('admin', 0).title).to.equal('Fred');
    expect(resource.link('admin', 1).href).to.equal('/admins/5');
    expect(resource.link('admin', 1).title).to.equal('Kate');
  });

  it('should parse the standard HAL example - properties', function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);

    expect(resource.allLinkArrays()).to.exist;
    expect(resource.currentlyProcessing).to.equal(14);
    expect(resource.shippedToday).to.equal(20);
  });

  it('should parse the standard HAL example - embedded resources', function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);

    expect(resource.allLinkArrays()).to.exist;
    expect(resource.allEmbeddedResourceArrays()).to.exist;
    expect(resource.allEmbeddedResourceArrays()).to.be.an('object');
    var orders = resource.embeddedResourceArray('orders');
    expect(orders).to.exist;
    expect(orders).to.be.an('array');
    expect(orders.length).to.equal(2);
    var order1 = resource.embedded('orders');
    var order2 = resource.embedded('orders', 1);
    expect(orders[0]).to.deep.equal(order1);
    expect(orders[1]).to.deep.equal(order2);

    expect(order1.allLinkArrays()).to.exist;
    expect(order1.allLinkArrays()).to.be.an('object');
    expect(order1.allLinkArrays().self[0]).to.exist;
    expect(order1.allLinkArrays().self[0].href).to.exist;
    expect(order1.linkArray('self')).to.exist;
    expect(order1.linkArray('self')).to.be.an('array');
    expect(order1.linkArray('self')[0]).to.exist;
    expect(order1.linkArray('self')[0].href).to.exist;

    expect(order1.link('self').href).to.equal('/orders/123');
    expect(order1.link('self').href).to.equal('/orders/123');
    expect(order1.link('basket').href).to.equal('/baskets/98712');
    expect(order1.link('customer').href).to.equal('/customers/7809');
    expect(order1.total).to.equal(30);
    expect(order1.currency).to.equal('USD');
    expect(order1.status).to.equal('shipped');

    expect(order2.link('self').href).to.equal('/orders/124');
    expect(order2.link('basket').href).to.equal('/baskets/97213');
    expect(order2.link('customer').href).to.equal('/customers/12369');
    expect(order2.total).to.equal(20);
    expect(order2.currency).to.equal('USD');
    expect(order2.status).to.equal('processing');
  });

  it('should not modify the embedded resources in the source object',
      function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);

    expect(unparsed._links.self).to.not.be.an('array');
    expect(unparsed._links.self).to.be.an('object');
    expect(unparsed._links.self.templated).to.not.exist;
    expect(unparsed._links.next).to.not.be.an('array');
    expect(unparsed._links.next).to.be.an('object');
    expect(unparsed._links.next.templated).to.not.exist;
    expect(unparsed._links.admin).to.be.an('array');
    expect(unparsed._links.find.templated).to.be.true;
    expect(unparsed.currentlyProcessing).to.equal(14);
    expect(unparsed._embedded).to.exist;
    expect(unparsed._validation).to.not.exist;

    expect(unparsed._embedded.orders[0]._links.self).to.not.be.an('array');
    expect(unparsed._embedded.orders[0]._links.self).to.be.an('object');
    expect(unparsed._embedded.orders[0]._links.self.templated).to.not.exist;
    expect(unparsed._embedded.orders[1]._links.basket).to.not.be.an('array');
    expect(unparsed._embedded.orders[1]._links.basket).to.be.an('object');
    expect(unparsed._embedded.orders[1]._links.basket.templated).to.not.exist;
    expect(unparsed._embedded.orders[0]._embedded).to.not.exist;
    expect(unparsed._embedded.orders[0]._validation).to.not.exist;
    expect(unparsed._embedded.orders[1]._embedded).to.not.exist;
    expect(unparsed._embedded.orders[1]._validation).to.not.exist;

    expect(resource._links.self).to.be.an('array');
    expect(resource._links.next).to.be.an('array');
    expect(resource._validation).to.exist;
    expect(resource._validation).to.be.an('array');

    expect(resource._embedded.orders[0]._links.self).to.be.an('array');
    expect(resource._embedded.orders[1]._links.basket).to.be.an('array');
    expect(resource._embedded.orders[0]._embedded).to.be.an('object');
    expect(resource._embedded.orders[0]._validation).to.be.an('array');
    expect(resource._embedded.orders[1]._embedded).to.be.an('object');
    expect(resource._embedded.orders[1]._validation).to.be.an('array');
  });

  it('should store the original source object in the resource',
      function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);
    expect(resource.original())
       .to.deep.equal(unparsed);
    expect(resource.embedded('orders', 1).original())
       .to.deep.equal(unparsed._embedded.orders[1]);
  });

  it('should parse curies', function() {
    var unparsed = fixtures.curies.get();
    var resource = halfred.parse(unparsed);
    expect(resource.hasCuries()).to.be.true;
    expect(resource.curieArray()).to.exist;
    expect(resource.curieArray()).to.be.an('array');
    expect(resource.curieArray().length).to.equal(2);
    var fullUrl1 = resource.curie('curie1');
    expect(fullUrl1.name).to.equal('curie1');
    expect(fullUrl1.templated).to.be.true;
    expect(fullUrl1.href).to.equal('http://docs.example.com/relations/{rel}');
    var fullUrl2 = resource.curie('curie2');
    expect(fullUrl2.name).to.equal('curie2');
    expect(fullUrl2.templated).to.be.false;
    expect(fullUrl2.href).to.equal('http://docs.example.com/relations/curie2');
  });

  it('should recognize that there are no curies', function() {
    var unparsed = fixtures.shop.get();
    var resource = halfred.parse(unparsed);
    expect(resource.hasCuries()).to.be.false;
    expect(resource.curieArray()).to.be.an('array');
    expect(resource.curieArray().length).to.equal(0);
    expect(resource.curie('whatever')).to.not.exist;
  });

  it('should reverse resolve non-templated curies', function() {
    var unparsed = fixtures.curies.get();
    var resource = halfred.parse(unparsed);
    var curie = resource.reverseResolveCurie(
      'http://docs.example.com/relations/curie2');
    expect(curie).to.equal('curie2');
  });

  it('should reverse resolve templated curies', function() {
    var unparsed = fixtures.curies.get();
    var resource = halfred.parse(unparsed);
    var curie = resource.reverseResolveCurie(
      'http://docs.example.com/relations/value');
    expect(curie).to.equal('curie1:value');
  });

  it('should reverse resolve to null/undefined if curie url does not ' +
      'exist', function() {
    var unparsed = fixtures.curies.get();
    var resource = halfred.parse(unparsed);
    var curie = resource.reverseResolveCurie(
      'http://docs.example.com/does/not/exist');
    expect(curie).to.not.exist;
  });

  it('should reverse resolve to null/undefined if there is no templated ' +
      ' match', function() {
    var unparsed = fixtures.curies.get();
    var resource = halfred.parse(unparsed);
    var curie = resource.reverseResolveCurie(
      'http://docs.example.com/relations/doesnotexist');
    expect(curie).to.not.exist;
  });

  it('should parse a resource without links', function() {
    var unparsed = fixtures.noLinks.get();
    var resource = halfred.parse(unparsed);

    // check that allLinkArrays() returns an empty object
    var allLinkArrays = resource.allLinkArrays();
    expect(allLinkArrays).to.exist;
    Object.keys(allLinkArrays).forEach(function(key) {
      assert.fail();
    });

    // check that linkArray('something')/link('something') returns
    // null/undefined instead of throwing an exception
    var nonExistingLinkArray = resource.linkArray('whatever');
    expect(nonExistingLink).to.not.exist;
    var nonExistingLink = resource.link('whatever');
    expect(nonExistingLink).to.not.exist;

    expect(resource.embedded('an_embedded_resource')).to.exist;
    expect(resource.property).to.equal('value');
  });

  it('should parse a resource without embedded resources', function() {
    var unparsed = fixtures.noEmbedded.get();
    var resource = halfred.parse(unparsed);

    // check that allEmbeddedArrays() returns an empty object
    var allEmbeddedArrays = resource.allEmbeddedArrays();
    expect(allEmbeddedArrays).to.exist;
    Object.keys(allEmbeddedArrays).forEach(function(key) {
      assert.fail();
    });

    // check that embeddedResourceArray('something')/
    // embeddedResource('something') returns null/undefined instead of throwing
    // an exception
    var nonExistingEmbeddedArray = resource.embeddedResourceArray('whatever');
    expect(nonExistingEmbeddedArray).to.not.exist;
    var nonExistingEmbeddedResource = resource.embeddedResource('whatever');
    expect(nonExistingEmbeddedResource).to.not.exist;

    expect(resource.link('self')).to.exist;
    expect(resource.property).to.equal('value');
  });

  it('should not report validation issues when validation is turned off',
      function() {
    var unparsed = fixtures.validation.get();
    var resource = halfred.parse(unparsed);
    checkInvalidResource(resource);
    expect(resource.validationIssues()).to.deep.equal([]);
    expect(resource.embedded('one').validationIssues()).to.deep.equal([]);
    expect(resource.embedded('two').validationIssues()).to.deep.equal([]);
  });

  it('should report validation issues when validation is turned on',
      function() {
    halfred.enableValidation();
    var unparsed = fixtures.validation.get();
    var resource = halfred.parse(unparsed);

    checkInvalidResource(resource);

    var validationRoot = resource.validationIssues();
    expect(validationRoot).to.be.an('array');
    expect(validationRoot.length).to.equal(5);
    expect(validationRoot[0].path).to.equal('$._links.first');
    expect(validationRoot[0].message).to.equal(
        'Link misses required property href.');
    expect(validationRoot[1].path).to.equal('$._links.uri-template-1');
    expect(validationRoot[1].message).to.equal('Link seems to be an URI ' +
        'template but its "templated" property is not set to true.');
    expect(validationRoot[2].path).to.equal('$._links.uri-template-2');
    expect(validationRoot[2].message).to.equal('Link seems to be an URI ' +
        'template but its "templated" property is not set to true.');
    expect(validationRoot[3].path).to.equal('$._links.uri-template-3');
    expect(validationRoot[3].message).to.equal('Link seems to be an URI ' +
        'template but its "templated" property is not set to true.');
    expect(validationRoot[4].path).to.equal('$._links');
    expect(validationRoot[4].message).to.equal('Resource does not have a ' +
        'self link');

    var validationOne = resource.embeddedResource('one').validationIssues();
    expect(validationOne).to.be.an('array');
    expect(validationOne.length).to.equal(1);
    expect(validationOne[0].path).to.equal('$._embedded.one._links');
    expect(validationOne[0].message).to.equal('Resource does not have a self ' +
        'link');

    var validationTwo = resource.embeddedResource('two').validationIssues();
    expect(validationTwo).to.be.an('array');
    expect(validationTwo.length).to.equal(0);
    /*
    TODO Implement validation for URIs/URI template syntax
    expect(validationTwo.length).to.equal(1)
    expect(validationTwo[0].path).to.equal('$._embedded.two._links.self')
    expect(validationTwo[0].message).to.equal('Attribute href is not a valid ' +
        'URI or URI template.')
    */
  });

  it('should report deprecation to log', function() {
    var unparsed = fixtures.deprecation.get();
    var resource = halfred.parse(unparsed);
    expect(recordedLogs.warn.length).to.equal(1);
    expect(recordedLogs.warn[0]).to.equal('Link ' +
      '$._embedded.one._embedded.two._links.foo is deprecated, see ' +
      'http://api.io/deprecated/link/explanation');
  });

  it('should report primitive link objects correctly', function() {
    var unparsed = fixtures.primitiveLink.get();
    try {
      halfred.parse(unparsed);
      assert.fail();
    } catch (e) {
      expect(e.message).to.equal('Link object is not an actual object: ' +
        '/links-must-be-objects [string]');
    }
  });

  function checkInvalidResource(resource) {
    expect(resource.allLinkArrays()).to.exist;
    expect(resource.link('self')).to.not.exist;
    expect(resource.link('first')).to.exist;
    expect(resource.link('first').href).to.not.exist;
    expect(resource.link('second').type).to.exist;
    expect(resource.link('second').type).to.be.an('array');
    expect(resource.link('third').type).to.be.a('null');
    expect(resource.link('third').type).not.to.be.an('undefined');
  }
});


test/fixtures.js
'use strict';

module.exports = {
  curies: require('./fixtures/curies'),
  deprecation: require('./fixtures/deprecation'),
  minimal: require('./fixtures/minimal'),
  noEmbedded: require('./fixtures/no_embedded'),
  noLinks: require('./fixtures/no_links'),
  primitiveLink: require('./fixtures/primitive_link'),
  shop: require('./fixtures/shop'),
  validation: require('./fixtures/validation')
};


browser/test/suite.js
require('../../test/parser');
require('../../test/immutable_stack');


Gruntfile.js
'use strict';

module.exports = function(grunt) {

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    jshint: {
      files: [
        '**/*.js',
        'Gruntfile.js',
        '!node_modules/**/*',
        '!browser/dist/**/*',
        '!browser/example/assets/*.js',
        '!browser/test/browserified_tests.js'
      ],
      options: {
        jshintrc: '.jshintrc'
      }
    },

    mochaTest: {
      test: {
        options: {
          reporter: 'spec'
        },
        src: ['test/**/*.js']
      }
    },

    // remove all previous browserified builds
    clean: {
      dist: ['./browser/dist/**/*.js'],
      tests: ['./browser/test/browserified_tests.js']
    },

    // browserify the module
    browserify: {
      // This browserify build be used by users of the module. It contains a
      // UMD (universal module definition) and can be used via an AMD module
      // loader like RequireJS or by simply placing a script tag in the page,
      // which registers the module as a global var.
      standalone: {
        src: [ '<%= pkg.name %>.js' ],
        dest: './browser/dist/<%= pkg.name %>.js',
        options: {
          browserifyOptions: {
            standalone: '<%= pkg.name %>'
          }
        }
      },
      // This browserify build can be required by other browserify modules that
      // have been created with an --external parameter.
      external: {
        src: [ '<%= pkg.name %>.js' ],
        dest: './browser/dist/<%= pkg.name %>.external.js',
        options: {
          alias: [ './halfred.js:halfred' ]
        }
      },
      // Browserified tests.
      tests: {
        src: [ 'test/browser_suite.js' ],
        dest: './browser/test/browserified_tests.js',
        options: {
          external: [ './halfred.js:halfred' ],
          browserifyOptions: {
            // Embed source map for tests
            debug: true,
          }
        }
      }
    },

    // Uglify browser libs
    uglify: {
      dist: {
        files: {
          'browser/dist/<%= pkg.name %>.min.js':
              ['<%= browserify.standalone.dest %>'],
          'browser/dist/<%= pkg.name %>.external.min.js':
              ['<%= browserify.external.dest %>']
        }
      }
    },

    // run the mocha tests in PhantomJS
    mocha: {
      test: {
        src: [ 'browser/test/index.html' ],
        options: {
          timeout: 20000,
          reporter: 'spec',
        }
      }
    },

    watch: {
      files: ['<%= jshint.files %>', '**/*.json'],
      tasks: ['default']
    },
  });

  // load all grunt-tasks
  require('load-grunt-tasks')(grunt);

  grunt.registerTask('test', [
      'jshint',
      'mochaTest'
  ]);

  grunt.registerTask('build', [
    'clean',
    'browserify',
    'uglify'
  ]);

  grunt.registerTask('test-browser', [
    'mocha'
  ]);

  grunt.registerTask('default', [
    'test',
    'build',
    'test-browser'
  ]);
};


lib/immutable_stack.js
'use strict';

/*
 * A very naive copy-on-write immutable stack. Since the size of the stack
 * is equal to the depth of the embedded resources for one HAL resource, the bad
 * performance for the copy-on-write approach is probably not a problem at all.
 * Might be replaced by a smarter solution later. Or not. Whatever.
 */
function ImmutableStack() {
  if (arguments.length >= 1) {
    this._array = arguments[0];
  } else {
    this._array = [];
  }
}

ImmutableStack.prototype.array = function() {
  return this._array;
};

ImmutableStack.prototype.isEmpty = function(array) {
  return this._array.length === 0;
};

ImmutableStack.prototype.push = function(element) {
  var array = this._array.slice(0);
  array.push(element);
  return new ImmutableStack(array);
};

ImmutableStack.prototype.pop = function() {
  var array = this._array.slice(0, this._array.length - 1);
  return new ImmutableStack(array);
};

ImmutableStack.prototype.peek = function() {
  if (this.isEmpty()) {
    throw new Error('can\'t peek on empty stack');
  }
  return this._array[this._array.length - 1];
};

module.exports = ImmutableStack;


browser/example/halfred-examples.js
/* global halfred: true */
/* jshint -W101 */
(function() {
  'use strict';

  var rootUri = '/';

  function parseAll() {
    ['#minimal-result',
     '#another-result'].forEach(function(div) {
      $(div).html('<img src="assets/spinner.gif"/>');
    });
    parseMinimal();
    parseAnother();
  }

  var minimalSource = {
    _links: {
      self: {
        href: '/minimal'
      }
    },
    attribute: 'value'
  };

  var anotherSource = {
    _links: {
      self: { href: '/orders' },
      next: { href: '/orders?page=2' },
      find: { href: '/orders{?id}', templated: true },
      admin: [
        { href: '/admins/2', title: 'Fred' },
        { href: '/admins/5', title: 'Kate' }
      ]
    },
    currentlyProcessing: 14,
    shippedToday: 20,
    _embedded: {
      orders: [{
        _links: {
          self: { href: '/orders/123' },
          basket: { href: '/baskets/98712' },
          customer: { href: '/customers/7809' }
        },
        total: 30.00,
        currency: 'USD',
        status: 'shipped'
      },{
        _links: {
          self: { href: '/orders/124' },
          basket: { href: '/baskets/97213' },
          customer: { href: '/customers/12369' }
        },
        total: 20.00,
        currency: 'USD',
        status: 'processing'
      }]
    }
  };

  function parseMinimal() {
    $('#minimal-result').html('<img src="assets/spinner.gif"/>');
    var resource = halfred.parse(minimalSource);
    $('#minimal-result').html(
        'var resource = halfred.parse(halObject);\n\n' +
        'resource.link(\'self\'):\n' +
        JSON.stringify(resource.link('self'), null, 2) + '\n\n' +
        'resource.attribute: ' + resource.attribute
    );
  }

  function parseAnother() {
    $('#another-result').html('<img src="assets/spinner.gif"/>');
    var resource = halfred.parse(anotherSource);
    $('#another-result').html(
        'var resource = halfred.parse(halObject);\n\n' +
        'resource.link(\'self\'):\n' +
        JSON.stringify(resource.link('self'), null, 2) + '\n\n' +
        'resource.link(\'next\'):\n' +
        JSON.stringify(resource.link('next'), null, 2) + '\n\n' +
        'resource.link(\'find\'):\n' +
        JSON.stringify(resource.link('find'), null, 2) + '\n\n' +
        'resource.linkArray(\'admin\')[0]:\n' +
        JSON.stringify(resource.linkArray('admin')[0], null, 2) + '\n\n' +
        'resource.linkArray(\'admin\')[1]:\n' +
        JSON.stringify(resource.linkArray('admin')[1], null, 2) + '\n\n' +
        'resource.currentlyProcessing: ' + resource.currentlyProcessing + '\n\n' +
        'resource.shippedToday: ' + resource.shippedToday + '\n\n'+
        'resource.embeddedResourceArray(\'orders\')[0].link(\'self\'):\n' +
        JSON.stringify(resource.embeddedResourceArray('orders')[0].link('self'), null, 2) + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[0].link(\'basket\'):\n' +
        JSON.stringify(resource.embeddedResourceArray('orders')[0].link('basket'), null, 2) + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[0].link(\'customer\'):\n' +
        JSON.stringify(resource.embeddedResourceArray('orders')[0].link('customer'), null, 2) + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[0].total: ' + resource.embeddedResourceArray('orders')[0].total + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[0].currency: ' + resource.embeddedResourceArray('orders')[0].currency + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[0].status: ' + resource.embeddedResourceArray('orders')[0].status + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[1].link(\'self\'):\n' +
        JSON.stringify(resource.embeddedResourceArray('orders')[1].link('self'), null, 2) + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[1].link(\'basket\'):\n' +
        JSON.stringify(resource.embeddedResourceArray('orders')[1].link('basket'), null, 2) + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[1].link(\'customer\'):\n' +
        JSON.stringify(resource.embeddedResourceArray('orders')[1].link('customer'), null, 2) + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[1].total: ' + resource.embeddedResourceArray('orders')[1].total + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[1].currency: ' + resource.embeddedResourceArray('orders')[1].currency + '\n\n' +
        'resource.embeddedResourceArray(\'orders\')[1].status: ' + resource.embeddedResourceArray('orders')[1].status + '\n\n'
    );
  }

  $(document).ready(function () {
    $('#btn-all').on('click', parseAll);
    $('#btn-minimal').on('click', parseMinimal);
    $('#btn-another').on('click', parseAnother);

    $('#minimal-source').html('<pre>\n' +
      JSON.stringify(minimalSource, null, 2) +
      '\n</pre>');

    $('#another-source').html('<pre>\n' +
      JSON.stringify(anotherSource, null, 2) +
      '\n</pre>');
  });
})();
/* jshint +W101 */
